R/ingest.R

```diff
@@ -0,0 +1,252 @@
+{
+    opts <- list(...)
+
+    if(is.data.frame(src))
+    {
+        con <- textConnection(NULL, "w")
+        on.exit(close(con))
+        utils::write.table(src, con, row.names=FALSE, col.names=FALSE, sep=",")
+        body <- textConnectionValue(con)
+        opts <- utils::modifyList(opts, list(streamFormat="Csv"))
+    }
+    else body <- readLines(src)
+
+    ingest_uri <- httr::parse_url(database$server)
+    ingest_uri$path <- file.path("v1/rest/ingest", database$database, dest_table)
+    ingest_uri$query <- opts
+
+    headers <- httr::add_headers(Authorization=paste("Bearer", validate_token(ingestion_token)))
+
+    res <- httr::POST(ingest_uri, headers, body=body, encode="raw")
+
+    http_status_handler <- match.arg(http_status_handler)
+    if(http_status_handler == "pass")
+        return(res)
+
+    cont <- httr::content(res, simplifyVector=TRUE)
+    handler <- get(paste0(http_status_handler, "_for_status"), getNamespace("httr"))
+    handler(res, make_error_message(cont))
+    cont
+}
+
+
+ingest_indirect <- function(database, src, dest_table, staging_container=NULL, ...)
+{
+    if(!requireNamespace("AzureStor"))
+        stop("AzureStor package must be installed to do indirect ingestion", call.=FALSE)
+
+    opts <- utils::modifyList(list(...), list(
+        key=staging_container$endpoint$key,
+        token=staging_container$endpoint$token,
+        sas=staging_container$endpoint$sas))
+
+    if(is.data.frame(src))
+    {
+        utils::write.table(src, textConnection("con", "w"), row.names=FALSE, col.names=FALSE, sep=",")
+        src <- textConnection(con, "r")
+        opts <- utils::modifyList(opts, list(streamFormat="Csv"))
+    }
+
+    if(inherits(staging_container, "blob_container"))
+    {
+        uploadfunc <- get("upload_blob", getNamespace("AzureStor"))
+        uploadfunc(staging_container, src, dest_table)
+
+        url <- httr::parse_url(staging_container$endpoint$url)
+        url$path <- file.path(staging_container$name, dest_table)
+
+        do.call(ingest_blob, c(list(database, httr::build_url(url), dest_table), opts))
+    }
+    else if(inherits(staging_container, "adls_filesystem"))
+    {
+        uploadfunc <- get("upload_adls_file", getNamespace("AzureStor"))
+        uploadfunc(staging_container, src, dest_table)
+
+        url <- httr::parse_url(staging_container$endpoint$url)
+        url$scheme <- "abfss"
+        url$username <- staging_container$name
+        url$path <- dest_table
+
+        do.call(ingest_adls2, c(list(database, httr::build_url(url), dest_table), opts))
+    }
+    else stop("Unsupported staging container type", call.=FALSE)
+}
+
+
+ingest_inline <- function(database, src, dest_table, ...)
+{
+    prop_list <- get_ingestion_properties(...)
+
+    if(is.data.frame(src))
+    {
+        con <- textConnection(NULL, "w")
+        on.exit(close(con))
+        utils::write.csv(src, con, row.names=FALSE, col.names=FALSE)
```

Comment:
`write.csv` is still writing the column names even with `col.names=FALSE`, any idea why?

```r
> con <- textConnection(NULL, "w")
> src <- iris
> utils::write.csv(src, con, row.names=FALSE, col.names=FALSE)
Warning message:
In utils::write.csv(src, con, row.names = FALSE, col.names = FALSE) :
  attempt to set 'col.names' ignored
> con
A connection with                            
description "NULL"          
class       "textConnection"
mode        "w"             
text        "text"          
opened      "opened"        
can read    "no"            
can write   "yes"           
> records <- textConnectionValue(con)
> records
  [1] "\"Sepal.Length\",\"Sepal.Width\",\"Petal.Length\",\"Petal.Width\",\"Species\""
  [2] "5.1,3.5,1.4,0.2,\"setosa\""  
```

Reply:
That's just write.csv being annoying; you have to use write.table. If you look at ?write.csv, you'll see that some args are ignored.